modules/core/src/lib/widget-manager.ts

```diff
@@ -0,0 +1,216 @@
+import type {MjolnirPointerEvent, MjolnirGestureEvent} from 'mjolnir.js';
+import type Layer from './layer';
+
+import {EVENTS} from './constants';
+
+export interface Widget<PropsT = any> {
+  // / Populated by core when mounted
+  _element?: HTMLDivElement | null;
+  _viewId?: string | null;
+
+  // / Lifecycle hooks
+  /** Called when the widget is added to a Deck instance.
+   * @returns an optional UI element that should be appended to the Deck container */
+  onAdd: (params: {
+    /** The Deck instance that the widget is attached to */
+    deck: Deck;
+    /** The id of the view that the widget is attached to */
+    viewId: string | null;
+  }) => HTMLDivElement | null;
+  /** Called when the widget is removed */
+  onRemove: () => void;
+  /** Called to update widget options */
+  setProps?: (props: Partial<PropsT>) => void;
+
+  // / Optional event hooks
+  /** Called when the containing view is changed */
+  onViewportChange?: (viewport: Viewport) => void;
+  /** Called when the containing view is redrawn */
+  onRedraw?: (params: {viewports: Viewport[]; layers: Layer[]}) => void;
+  /** Called when a hover event occurs */
+  onHover?: (info: PickingInfo, event: MjolnirPointerEvent) => void;
+  /** Called when a click event occurs */
+  onClick?: (info: PickingInfo, event: MjolnirGestureEvent) => void;
+  /** Called when a drag event occurs */
+  onDrag?: (info: PickingInfo, event: MjolnirGestureEvent) => void;
+  /** Called when a dragstart event occurs */
+  onDragStart?: (info: PickingInfo, event: MjolnirGestureEvent) => void;
+  /** Called when a dragend event occurs */
+  onDragEnd?: (info: PickingInfo, event: MjolnirGestureEvent) => void;
+}
+
+const PLACEMENTS = {
+  'top-left': {top: 0, left: 0},
+  'top-right': {top: 0, right: 0},
+  'bottom-left': {bottom: 0, left: 0},
+  'bottom-right': {bottom: 0, right: 0},
+  fill: {top: 0, left: 0, bottom: 0, right: 0}
+} as const;
+
+export type WidgetPlacement = keyof typeof PLACEMENTS;
+
+const ROOT_CONTAINER_ID = '__root';
+
+export class WidgetManager {
+  deck: Deck;
+  parent?: HTMLElement | null;
+  containers: {[id: string]: HTMLDivElement} = {};
+  widgets: Widget[] = [];
+  lastViewports: {[id: string]: Viewport} = {};
+
+  constructor({deck, parent}: {deck: Deck; parent?: HTMLElement | null}) {
+    this.deck = deck;
+    this.parent = parent;
+  }
+
+  finalize() {
+    for (const widget of this.widgets) {
+      this.remove(widget);
+    }
+    for (const id in this.containers) {
+      this.containers[id].remove();
+    }
+  }
+
+  add(
+    widget: Widget,
+    opts: {
+      viewId?: string | null;
+      placement?: WidgetPlacement;
+    } = {}
+  ) {
+    if (this.widgets.includes(widget)) {
+      // widget already added
+      return;
+    }
+
+    const {placement = 'top-left', viewId = null} = opts;
+    const element = widget.onAdd({deck: this.deck, viewId});
+
+    if (element) {
+      this._getContainer(viewId, placement).append(element);
```

Comment:
It looks like widgets are added sequentially to the container, correct? Is there an easy way to re-order them in the DOM or do you have to ensure that they're added in the right order?

Reply:
We could support reorder in the declarative API (to be implemented). What is the use case you have in mind?